<a href="https://colab.research.google.com/github/Aidas-Baublys/Python-hello-world/blob/master/03_Pandas/PP3_Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# PP3: Pandas

## Section 1: Getting and knowing your data

### Step 1. Import the necessary libraries

In [64]:
# Run all cells to avoid simple errors like "not defined", because pandas are imported only once here.
# Also, colab sometimes does not recognise data frame type after it is assinged to variable with pd.read_csv
# so intelisense is incomplete and code blocks show false positive errors.
# Now we both know what to ingnore
import pandas as pd

### Step 2. Import the dataset from this [address](https://raw.githubusercontent.com/justmarkham/DAT8/master/data/u.user). 

In [2]:
data_url = "https://raw.githubusercontent.com/justmarkham/DAT8/master/data/u.user"

### Step 3. Assign it to a variable called users and use the 'user_id' as index

In [3]:
users = pd.read_csv(data_url, sep="|", index_col="user_id")

### Step 4. See the first 25 entries

In [4]:
users.head(25)

age gender     occupation zip_code
user_id                                    
1         24      M     technician    85711
2         53      F          other    94043
3         23      M         writer    32067
4         24      M     technician    43537
5         33      F          other    15213
6         42      M      executive    98101
7         57      M  administrator    91344
8         36      M  administrator    05201
9         29      M        student    01002
10        53      M         lawyer    90703
11        39      F          other    30329
12        28      F          other    06405
13        47      M       educator    29206
14        45      M      scientist    55106
15        49      F       educator    97301
16        21      M  entertainment    10309
17        30      M     programmer    06355
18        35      F          other    37212
19        40      M      librarian    02138
20        42      F      homemaker    95660
21        26      M         writer    30068
22        25      M         writer    40206
23        30      F         artist    48197
24        21      F         artist    94533
25        39      M       engineer    55107

### Step 5. See the last 10 entries

In [5]:
users.tail(10)

age gender     occupation zip_code
user_id                                    
934       61      M       engineer    22902
935       42      M         doctor    66221
936       24      M          other    32789
937       48      M       educator    98072
938       38      F     technician    55038
939       26      F        student    33319
940       32      M  administrator    02215
941       20      M        student    97229
942       48      F      librarian    78209
943       22      M        student    77841

### Step 6. What is the number of observations in the dataset?

In [6]:
# Observations are rows acording to this article https://medium.com/data-science-365/pandas-for-data-science-part-2-c12c3ee876c2, so
users.shape[0]

943

### Step 7. What is the number of columns in the dataset?

In [7]:
# Without user_id
users.shape[1]

4

In [8]:
# With user_id
5

5

### Step 8. Print the name of all the columns.

In [9]:
for col_name in users.columns.tolist():
  print(col_name)

age
gender
occupation
zip_code


### Step 9. How is the dataset indexed?

In [10]:
# Now by user_id, but originally the default way by row count starting from 0

### Step 10. What is the data type of each column?

In [11]:
users.dtypes

age            int64
gender        object
occupation    object
zip_code      object
dtype: object

### Step 11. Print only the occupation column

In [12]:
users.occupation

user_id
1         technician
2              other
3             writer
4         technician
5              other
           ...      
939          student
940    administrator
941          student
942        librarian
943          student
Name: occupation, Length: 943, dtype: object

### Step 12. How many different occupations are in this dataset?

In [13]:
users.occupation.unique()

array(['technician', 'other', 'writer', 'executive', 'administrator',
       'student', 'lawyer', 'educator', 'scientist', 'entertainment',
       'programmer', 'librarian', 'homemaker', 'artist', 'engineer',
       'marketing', 'none', 'healthcare', 'retired', 'salesman', 'doctor'],
      dtype=object)

In [14]:
# If we include "other"
len(users.occupation.unique())

21

### Step 13. What is the most frequent occupation?

In [15]:
# Studnet
users.occupation.value_counts()

student          196
other            105
educator          95
administrator     79
engineer          67
programmer        66
librarian         51
writer            45
executive         32
scientist         31
artist            28
technician        27
marketing         26
entertainment     18
healthcare        16
retired           14
lawyer            12
salesman          12
none               9
homemaker          7
doctor             7
Name: occupation, dtype: int64

### Step 14. Summarize the DataFrame.

In [16]:
# Basing answer on this https://stackoverflow.com/questions/24524104/pandas-describe-is-not-returning-summary-of-all-columns 
# and this https://www.youtube.com/watch?v=Q06Y3DUSwz4&ab_channel=DataScienceTutorials
# By default describe ignores non numeric values.
users.describe()

age
count  943.000000
mean    34.051962
std     12.192740
min      7.000000
25%     25.000000
50%     31.000000
75%     43.000000
max     73.000000

### Step 15. Summarize all the columns

In [17]:
users.describe(include="all")

age gender occupation zip_code
count   943.000000    943        943      943
unique         NaN      2         21      795
top            NaN      M    student    55414
freq           NaN    670        196        9
mean     34.051962    NaN        NaN      NaN
std      12.192740    NaN        NaN      NaN
min       7.000000    NaN        NaN      NaN
25%      25.000000    NaN        NaN      NaN
50%      31.000000    NaN        NaN      NaN
75%      43.000000    NaN        NaN      NaN
max      73.000000    NaN        NaN      NaN

### Step 16. Summarize only the occupation column

In [18]:
# Columns can also be accessed by dot notation
users["occupation"].describe()

count         943
unique         21
top       student
freq          196
Name: occupation, dtype: object

### Step 17. What is the mean age of users?

In [19]:
users.describe().loc[["mean"]]

age
mean  34.051962

### Step 18. What is the age with least occurrence?

In [20]:
users.age.value_counts()

30    39
25    38
22    37
28    36
27    35
      ..
7      1
66     1
11     1
10     1
73     1
Name: age, Length: 61, dtype: int64

In [21]:
# These five ages
users.age.value_counts().tail(5)

7     1
66    1
11    1
10    1
73    1
Name: age, dtype: int64

## Section 2: Filtering and Sorting

### Step 1. Import the necessary libraries

In [22]:
# Already imported at the top of notebook

### Step 2. Import the dataset from this [address](https://raw.githubusercontent.com/guipsamora/pandas_exercises/master/02_Filtering_%26_Sorting/Euro12/Euro_2012_stats_TEAM.csv). 

In [23]:
euro_data_url = "https://raw.githubusercontent.com/guipsamora/pandas_exercises/master/02_Filtering_%26_Sorting/Euro12/Euro_2012_stats_TEAM.csv"

### Step 3. Assign it to a variable called euro12.

In [24]:
euro12 = pd.read_csv(euro_data_url)

### Step 4. Select only the Goal column.

In [25]:
euro12["Goals"]

0      4
1      4
2      4
3      5
4      3
5     10
6      5
7      6
8      2
9      2
10     6
11     1
12     5
13    12
14     5
15     2
Name: Goals, dtype: int64

### Step 5. How many team participated in the Euro2012?

In [26]:
euro12["Team"].count()

16

### Step 6. What is the number of columns in the dataset?

In [27]:
euro12.shape[1]

35

### Step 7. View only the columns Team, Yellow Cards and Red Cards and assign them to a dataframe called discipline

In [28]:
discipline = euro12[["Team", "Yellow Cards", "Red Cards"]]
discipline

Team  Yellow Cards  Red Cards
0               Croatia             9          0
1        Czech Republic             7          0
2               Denmark             4          0
3               England             5          0
4                France             6          0
5               Germany             4          0
6                Greece             9          1
7                 Italy            16          0
8           Netherlands             5          0
9                Poland             7          1
10             Portugal            12          0
11  Republic of Ireland             6          1
12               Russia             6          0
13                Spain            11          0
14               Sweden             7          0
15              Ukraine             5          0

### Step 8. Sort the teams by Red Cards, then to Yellow Cards

In [29]:
discipline.sort_values("Red Cards")

Team  Yellow Cards  Red Cards
0               Croatia             9          0
1        Czech Republic             7          0
2               Denmark             4          0
3               England             5          0
4                France             6          0
5               Germany             4          0
7                 Italy            16          0
8           Netherlands             5          0
10             Portugal            12          0
12               Russia             6          0
13                Spain            11          0
14               Sweden             7          0
15              Ukraine             5          0
6                Greece             9          1
9                Poland             7          1
11  Republic of Ireland             6          1

In [30]:
discipline.sort_values("Yellow Cards")

Team  Yellow Cards  Red Cards
2               Denmark             4          0
5               Germany             4          0
3               England             5          0
8           Netherlands             5          0
15              Ukraine             5          0
4                France             6          0
11  Republic of Ireland             6          1
12               Russia             6          0
1        Czech Republic             7          0
9                Poland             7          1
14               Sweden             7          0
0               Croatia             9          0
6                Greece             9          1
13                Spain            11          0
10             Portugal            12          0
7                 Italy            16          0

### Step 9. Calculate the mean Yellow Cards given per Team

In [31]:
discipline["Yellow Cards"].mean()

7.4375

### Step 10. Filter teams that scored more than 6 goals

In [32]:
# Have to use euro12 dataframe, cause discipline does not have "Goals" column
a_lot_of_goals_filter = euro12["Goals"] > 6
euro12[a_lot_of_goals_filter]

Team  Goals  Shots on target  Shots off target Shooting Accuracy  \
5   Germany     10               32                32             47.8%   
13    Spain     12               42                33             55.9%   

   % Goals-to-shots  Total shots (inc. Blocked)  Hit Woodwork  Penalty goals  \
5             15.6%                          80             2              1   
13            16.0%                         100             0              1   

    Penalties not scored  ...  Saves made  Saves-to-shots ratio  Fouls Won  \
5                      0  ...          10                 62.6%         63   
13                     0  ...          15                 93.8%        102   

   Fouls Conceded  Offsides  Yellow Cards  Red Cards  Subs on  Subs off  \
5              49        12             4          0       15        15   
13             83        19            11          0       17        17   

    Players Used  
5             17  
13            18  

[2 rows x 35 columns]

### Step 11. Select the teams that start with G

In [33]:
team_g_selector = euro12["Team"].str.startswith("G")
discipline[team_g_selector]

Team  Yellow Cards  Red Cards
5  Germany             4          0
6   Greece             9          1

### Step 12. Select the first 7 columns

In [34]:
discipline.head(7)

Team  Yellow Cards  Red Cards
0         Croatia             9          0
1  Czech Republic             7          0
2         Denmark             4          0
3         England             5          0
4          France             6          0
5         Germany             4          0
6          Greece             9          1

### Step 13. Select all columns except the last 3.

In [35]:
discipline[:-3]

Team  Yellow Cards  Red Cards
0               Croatia             9          0
1        Czech Republic             7          0
2               Denmark             4          0
3               England             5          0
4                France             6          0
5               Germany             4          0
6                Greece             9          1
7                 Italy            16          0
8           Netherlands             5          0
9                Poland             7          1
10             Portugal            12          0
11  Republic of Ireland             6          1
12               Russia             6          0

### Step 14. Present only the Shooting Accuracy from England, Italy and Russia

In [36]:
countries = ["England", "Italy", "Russia"]
country_filter = euro12["Team"].isin(countries)
euro12.loc[country_filter, ["Team", "Shooting Accuracy"]]

Team Shooting Accuracy
3   England             50.0%
7     Italy             43.0%
12   Russia             22.5%

## Section 3: Grouping

### Step 1. Import the necessary libraries

In [37]:
# Already imported at the top of notebook

### Step 2. Import the dataset from this [address](https://raw.githubusercontent.com/justmarkham/DAT8/master/data/drinks.csv). 

In [38]:
drinks_url = "https://raw.githubusercontent.com/justmarkham/DAT8/master/data/drinks.csv"

### Step 3. Assign it to a variable called drinks.

In [39]:
drinks = pd.read_csv(drinks_url)

### Step 4. Which continent drinks more beer on average?

In [40]:
continent_group = drinks.groupby(["continent"])
sort_by_beer = continent_group["beer_servings"].mean().sort_values(ascending=False)
sort_by_beer.iloc[:1]

continent
EU    193.777778
Name: beer_servings, dtype: float64

### Step 5. For each continent print the statistics for wine consumption.

In [41]:
continent_group["wine_servings"].describe()

count        mean        std  min   25%    50%     75%    max
continent                                                               
AF          53.0   16.264151  38.846419  0.0   1.0    2.0   13.00  233.0
AS          44.0    9.068182  21.667034  0.0   0.0    1.0    8.00  123.0
EU          45.0  142.222222  97.421738  0.0  59.0  128.0  195.00  370.0
OC          16.0   35.625000  64.555790  0.0   1.0    8.5   23.25  212.0
SA          12.0   62.416667  88.620189  1.0   3.0   12.0   98.50  221.0

### Step 6. Print the mean alcohol consumption per continent for every column

In [42]:
# This works, but I get "FutureWarning: Indexing with multiple keys..." warning, which we can ignore in the present together
continent_group["beer_servings", "spirit_servings", "wine_servings", "total_litres_of_pure_alcohol"].mean()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  


beer_servings  spirit_servings  wine_servings  \
continent                                                  
AF             61.471698        16.339623      16.264151   
AS             37.045455        60.840909       9.068182   
EU            193.777778       132.555556     142.222222   
OC             89.687500        58.437500      35.625000   
SA            175.083333       114.750000      62.416667   

           total_litres_of_pure_alcohol  
continent                                
AF                             3.007547  
AS                             2.170455  
EU                             8.617778  
OC                             3.381250  
SA                             6.308333

### Step 7. Print the median alcohol consumption per continent for every column

In [43]:
# The same here. The future is full of warnings nowadays
continent_group["beer_servings", "spirit_servings", "wine_servings", "total_litres_of_pure_alcohol"].median()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  


beer_servings  spirit_servings  wine_servings  \
continent                                                  
AF                  32.0              3.0            2.0   
AS                  17.5             16.0            1.0   
EU                 219.0            122.0          128.0   
OC                  52.5             37.0            8.5   
SA                 162.5            108.5           12.0   

           total_litres_of_pure_alcohol  
continent                                
AF                                 2.30  
AS                                 1.20  
EU                                10.00  
OC                                 1.75  
SA                                 6.85

### Step 8. Print the mean, min and max values for spirit consumption.
#### This time output a DataFrame

In [44]:
# Me thinks I putted data frame type out all along, yes I did
some_stats = continent_group["spirit_servings"].describe()[["mean", "min", "max"]]
print(type(some_stats))
some_stats

<class 'pandas.core.frame.DataFrame'>


mean   min    max
continent                         
AF          16.339623   0.0  152.0
AS          60.840909   0.0  326.0
EU         132.555556   0.0  373.0
OC          58.437500   0.0  254.0
SA         114.750000  25.0  302.0

## Section 4: Apply

### Step 1. Import the necessary libraries

In [45]:
# You know already

### Step 2. Import the dataset from this [address](https://raw.githubusercontent.com/guipsamora/pandas_exercises/master/04_Apply/US_Crime_Rates/US_Crime_Rates_1960_2014.csv). 

In [46]:
crime_url = "https://raw.githubusercontent.com/guipsamora/pandas_exercises/master/04_Apply/US_Crime_Rates/US_Crime_Rates_1960_2014.csv"

### Step 3. Assign it to a variable called crime.

In [47]:
crime = pd.read_csv(crime_url)

### Step 4. What is the type of the columns?

In [48]:
crime.dtypes

Year                  int64
Population            int64
Total                 int64
Violent               int64
Property              int64
Murder                int64
Forcible_Rape         int64
Robbery               int64
Aggravated_assault    int64
Burglary              int64
Larceny_Theft         int64
Vehicle_Theft         int64
dtype: object

##### Have you noticed that the type of Year is int64. But pandas has a different type to work with Time Series. Let's see it now.

### Step 5. Convert the type of the column Year to datetime64

In [49]:
crime["Year"] = pd.to_datetime(crime["Year"], format="%Y")
crime.dtypes

Year                  datetime64[ns]
Population                     int64
Total                          int64
Violent                        int64
Property                       int64
Murder                         int64
Forcible_Rape                  int64
Robbery                        int64
Aggravated_assault             int64
Burglary                       int64
Larceny_Theft                  int64
Vehicle_Theft                  int64
dtype: object

### Step 6. Set the Year column as the index of the dataframe

In [50]:
# Will only set for this code block
crime.set_index("Year")

Population     Total  Violent  Property  Murder  Forcible_Rape  \
Year                                                                         
1960-01-01   179323175   3384200   288460   3095700    9110          17190   
1961-01-01   182992000   3488000   289390   3198600    8740          17220   
1962-01-01   185771000   3752200   301510   3450700    8530          17550   
1963-01-01   188483000   4109500   316970   3792500    8640          17650   
1964-01-01   191141000   4564600   364220   4200400    9360          21420   
1965-01-01   193526000   4739400   387390   4352000    9960          23410   
1966-01-01   195576000   5223500   430180   4793300   11040          25820   
1967-01-01   197457000   5903400   499930   5403500   12240          27620   
1968-01-01   199399000   6720200   595010   6125200   13800          31670   
1969-01-01   201385000   7410900   661870   6749000   14760          37170   
1970-01-01   203235298   8098000   738820   7359200   16000          37990   
1971-01-01   206212000   8588200   816500   7771700   17780          42260   
1972-01-01   208230000   8248800   834900   7413900   18670          46850   
1973-01-01   209851000   8718100   875910   7842200   19640          51400   
1974-01-01   211392000  10253400   974720   9278700   20710          55400   
1975-01-01   213124000  11292400  1039710  10252700   20510          56090   
1976-01-01   214659000  11349700  1004210  10345500   18780          57080   
1977-01-01   216332000  10984500  1029580   9955000   19120          63500   
1978-01-01   218059000  11209000  1085550  10123400   19560          67610   
1979-01-01   220099000  12249500  1208030  11041500   21460          76390   
1980-01-01   225349264  13408300  1344520  12063700   23040          82990   
1981-01-01   229146000  13423800  1361820  12061900   22520          82500   
1982-01-01   231534000  12974400  1322390  11652000   21010          78770   
1983-01-01   233981000  12108600  1258090  10850500   19310          78920   
1984-01-01   236158000  11881800  1273280  10608500   18690          84230   
1985-01-01   238740000  12431400  1328800  11102600   18980          88670   
1986-01-01   240132887  13211869  1489169  11722700   20613          91459   
1987-01-01   242282918  13508700  1483999  12024700   20096          91110   
1988-01-01   245807000  13923100  1566220  12356900   20680          92490   
1989-01-01   248239000  14251400  1646040  12605400   21500          94500   
1990-01-01   248709873  14475600  1820130  12655500   23440         102560   
1991-01-01   252177000  14872900  1911770  12961100   24700         106590   
1992-01-01   255082000  14438200  1932270  12505900   23760         109060   
1993-01-01   257908000  14144800  1926020  12218800   24530         106010   
1994-01-01   260341000  13989500  1857670  12131900   23330         102220   
1995-01-01   262755000  13862700  1798790  12063900   21610          97470   
1996-01-01   265228572  13493863  1688540  11805300   19650          96250   
1997-01-01   267637000  13194571  1634770  11558175   18208          96153   
1998-01-01   270296000  12475634  1531044  10944590   16914          93103   
1999-01-01   272690813  11634378  1426044  10208334   15522          89411   
2000-01-01   281421906  11608072  1425486  10182586   15586          90178   
2001-01-01   285317559  11876669  1439480  10437480   16037          90863   
2002-01-01   287973924  11878954  1423677  10455277   16229          95235   
2003-01-01   290690788  11826538  1383676  10442862   16528          93883   
2004-01-01   293656842  11679474  1360088  10319386   16148          95089   
2005-01-01   296507061  11565499  1390745  10174754   16740          94347   
2006-01-01   299398484  11401511  1418043   9983568   17030          92757   
2007-01-01   301621157  11251828  1408337   9843481   16929          90427   
2008-01-01   304374846  11160543  1392628   9767915   16442          90479   
2009-01-01   307006550  10762956  

### Step 7. Delete the Total column

In [51]:
#  Will only delete for this code block
crime.drop(columns="Total")

Year  Population  Violent  Property  Murder  Forcible_Rape  Robbery  \
0  1960-01-01   179323175   288460   3095700    9110          17190   107840   
1  1961-01-01   182992000   289390   3198600    8740          17220   106670   
2  1962-01-01   185771000   301510   3450700    8530          17550   110860   
3  1963-01-01   188483000   316970   3792500    8640          17650   116470   
4  1964-01-01   191141000   364220   4200400    9360          21420   130390   
5  1965-01-01   193526000   387390   4352000    9960          23410   138690   
6  1966-01-01   195576000   430180   4793300   11040          25820   157990   
7  1967-01-01   197457000   499930   5403500   12240          27620   202910   
8  1968-01-01   199399000   595010   6125200   13800          31670   262840   
9  1969-01-01   201385000   661870   6749000   14760          37170   298850   
10 1970-01-01   203235298   738820   7359200   16000          37990   349860   
11 1971-01-01   206212000   816500   7771700   17780          42260   387700   
12 1972-01-01   208230000   834900   7413900   18670          46850   376290   
13 1973-01-01   209851000   875910   7842200   19640          51400   384220   
14 1974-01-01   211392000   974720   9278700   20710          55400   442400   
15 1975-01-01   213124000  1039710  10252700   20510          56090   470500   
16 1976-01-01   214659000  1004210  10345500   18780          57080   427810   
17 1977-01-01   216332000  1029580   9955000   19120          63500   412610   
18 1978-01-01   218059000  1085550  10123400   19560          67610   426930   
19 1979-01-01   220099000  1208030  11041500   21460          76390   480700   
20 1980-01-01   225349264  1344520  12063700   23040          82990   565840   
21 1981-01-01   229146000  1361820  12061900   22520          82500   592910   
22 1982-01-01   231534000  1322390  11652000   21010          78770   553130   
23 1983-01-01   233981000  1258090  10850500   19310          78920   506570   
24 1984-01-01   236158000  1273280  10608500   18690          84230   485010   
25 1985-01-01   238740000  1328800  11102600   18980          88670   497870   
26 1986-01-01   240132887  1489169  11722700   20613          91459   542775   
27 1987-01-01   242282918  1483999  12024700   20096          91110   517704   
28 1988-01-01   245807000  1566220  12356900   20680          92490   542970   
29 1989-01-01   248239000  1646040  12605400   21500          94500   578330   
30 1990-01-01   248709873  1820130  12655500   23440         102560   639270   
31 1991-01-01   252177000  1911770  12961100   24700         106590   687730   
32 1992-01-01   255082000  1932270  12505900   23760         109060   672480   
33 1993-01-01   257908000  1926020  12218800   24530         106010   659870   
34 1994-01-01   260341000  1857670  12131900   23330         102220   618950   
35 1995-01-01   262755000  1798790  12063900   21610          97470   580510   
36 1996-01-01   265228572  1688540  11805300   19650          96250   535590   
37 1997-01-01   267637000  1634770  11558175   18208          96153   498534   
38 1998-01-01   270296000  1531044  10944590   16914          93103   446625   
39 1999-01-01   272690813  1426044  10208334   15522          89411   409371   
40 2000-01-01   281421906  1425486  10182586   15586          90178   408016   
41 2001-01-01   285317559  1439480  10437480   16037          90863   423557   
42 2002-01-01   287973924  1423677  10455277   16229          95235   420806   
43 2003-01-01   290690788  1383676  10442862   16528          93883   414235   
44 2004-01-01   293656842  1360088  10319386   16148          95089   401470   
45 2005-01-01   296507061  1390745  10174754   16740          94347   417438   
46 2006-01-01   299398484  1418043   9983568   17030          92757   447403   
47 2007-01-01   301621157  1408337   9843481   16929          90427   445125   
48 2008-01-01   304374846  1392628   9767915   16442          90479   443574   
49 2009-0

### Step 8. Group the year by decades and sum the values

#### Pay attention to the Population column number, summing this column is a mistake

In [52]:
# Dropping the mistake, setting year as index for resample to work and summing
crime_by_decade = crime.drop(columns="Population").set_index('Year').resample('10AS').sum()
crime_by_decade

Total   Violent   Property  Murder  Forcible_Rape  Robbery  \
Year                                                                         
1960-01-01   49295900   4134930   45160900  106180         236720  1633510   
1970-01-01  100991600   9607930   91383800  192230         554570  4159020   
1980-01-01  131123369  14074328  117048900  206439         865639  5383109   
1990-01-01  136582146  17527048  119053499  211664         998827  5748930   
2000-01-01  115012044  13968056  100944369  163068         922499  4230366   
2010-01-01   50167967   6072017   44095950   72867         421059  1749809   

            Aggravated_assault  Burglary  Larceny_Theft  Vehicle_Theft  
Year                                                                    
1960-01-01             2158520  13321100       26547700        5292100  
1970-01-01             4702120  28486000       53157800        9739900  
1980-01-01             7619130  33073494       72040253       11935411  
1990-01-01            10568963  26750015       77679366       14624418  
2000-01-01             8652124  21565176       67970291       11412834  
2010-01-01             3764142  10125170       30401698        3569080

### Step 9. What is the most dangerous decade to live in the US?

In [53]:
# The nineties, man, by all accounts and metrics
crime_categories = ["Total", "Violent",	"Property",	"Murder", "Forcible_Rape", "Robbery", "Aggravated_assault",	"Burglary",	"Larceny_Theft", "Vehicle_Theft"]

for category in crime_categories:
  print(crime_by_decade.sort_values(category, ascending=False)[category].head(1))

Year
1990-01-01    136582146
Name: Total, dtype: int64
Year
1990-01-01    17527048
Name: Violent, dtype: int64
Year
1990-01-01    119053499
Name: Property, dtype: int64
Year
1990-01-01    211664
Name: Murder, dtype: int64
Year
1990-01-01    998827
Name: Forcible_Rape, dtype: int64
Year
1990-01-01    5748930
Name: Robbery, dtype: int64
Year
1990-01-01    10568963
Name: Aggravated_assault, dtype: int64
Year
1980-01-01    33073494
Name: Burglary, dtype: int64
Year
1990-01-01    77679366
Name: Larceny_Theft, dtype: int64
Year
1990-01-01    14624418
Name: Vehicle_Theft, dtype: int64


## Section 5: Merge

### Step 1. Import the necessary libraries

In [54]:
# Yep.

### Step 2. Create the 3 DataFrames based on the following raw data

In [55]:
raw_data_1 = {
        'subject_id': ['1', '2', '3', '4', '5'],
        'first_name': ['Alex', 'Amy', 'Allen', 'Alice', 'Ayoung'], 
        'last_name': ['Anderson', 'Ackerman', 'Ali', 'Aoni', 'Atiches']}

raw_data_2 = {
        'subject_id': ['4', '5', '6', '7', '8'],
        'first_name': ['Billy', 'Brian', 'Bran', 'Bryce', 'Betty'], 
        'last_name': ['Bonder', 'Black', 'Balwner', 'Brice', 'Btisan']}

raw_data_3 = {
        'subject_id': ['1', '2', '3', '4', '5', '7', '8', '9', '10', '11'],
        'test_id': [51, 15, 15, 61, 16, 14, 15, 1, 61, 16]}

### Step 3. Assign each to a variable called data1, data2, data3

In [56]:
data1 = pd.DataFrame(raw_data_1)
data2 = pd.DataFrame(raw_data_2)
data3 = pd.DataFrame(raw_data_3)

### Step 4. Join the two dataframes along rows and assign all_data

In [57]:
# Assuming data1 and data2 need to be joined and that all_data is a variable
all_data = pd.concat([data1, data2])
all_data

subject_id first_name last_name
0          1       Alex  Anderson
1          2        Amy  Ackerman
2          3      Allen       Ali
3          4      Alice      Aoni
4          5     Ayoung   Atiches
0          4      Billy    Bonder
1          5      Brian     Black
2          6       Bran   Balwner
3          7      Bryce     Brice
4          8      Betty    Btisan

### Step 5. Join the two dataframes along columns and assing to all_data_col

In [58]:
# This question is also incomplete, so choosing subject_id column
all_data_col = pd.merge(data1, data2, on='subject_id', how='outer')

### Step 6. Print data3

In [59]:
data3

subject_id  test_id
0          1       51
1          2       15
2          3       15
3          4       61
4          5       16
5          7       14
6          8       15
7          9        1
8         10       61
9         11       16

### Step 7. Merge all_data and data3 along the subject_id value

In [60]:
pd.merge(all_data, data3, on='subject_id', how='outer')

subject_id first_name last_name  test_id
0           1       Alex  Anderson     51.0
1           2        Amy  Ackerman     15.0
2           3      Allen       Ali     15.0
3           4      Alice      Aoni     61.0
4           4      Billy    Bonder     61.0
5           5     Ayoung   Atiches     16.0
6           5      Brian     Black     16.0
7           6       Bran   Balwner      NaN
8           7      Bryce     Brice     14.0
9           8      Betty    Btisan     15.0
10          9        NaN       NaN      1.0
11         10        NaN       NaN     61.0
12         11        NaN       NaN     16.0

### Step 8. Merge only the data that has the same 'subject_id' on both data1 and data2

In [61]:
pd.merge(data1, data2, on="subject_id")

subject_id first_name_x last_name_x first_name_y last_name_y
0          4        Alice        Aoni        Billy      Bonder
1          5       Ayoung     Atiches        Brian       Black

### Step 9. Merge all values in data1 and data2, with matching records from both sides where available.

In [90]:
# Not sure I understood the question. Leave matching values? And matching as in equal?
pd.merge(data1, data2, how="outer") 

subject_id first_name last_name
0          1       Alex  Anderson
1          2        Amy  Ackerman
2          3      Allen       Ali
3          4      Alice      Aoni
4          5     Ayoung   Atiches
5          4      Billy    Bonder
6          5      Brian     Black
7          6       Bran   Balwner
8          7      Bryce     Brice
9          8      Betty    Btisan

## Section 6: Stats

### Introduction:

The data have been modified to contain some missing values, identified by NaN.  
Using pandas should make this exercise
easier, in particular for the bonus question.

You should be able to perform all of these operations without using
a for loop or other looping construct.


1. The data in 'wind.data' has the following format:

In [91]:
"""
Yr Mo Dy   RPT   VAL   ROS   KIL   SHA   BIR   DUB   CLA   MUL   CLO   BEL   MAL
61  1  1 15.04 14.96 13.17  9.29   NaN  9.87 13.67 10.25 10.83 12.58 18.50 15.04
61  1  2 14.71   NaN 10.83  6.50 12.62  7.67 11.50 10.04  9.79  9.67 17.54 13.83
61  1  3 18.50 16.88 12.33 10.13 11.17  6.17 11.25   NaN  8.50  7.67 12.75 12.71
"""

'\nYr Mo Dy   RPT   VAL   ROS   KIL   SHA   BIR   DUB   CLA   MUL   CLO   BEL   MAL\n61  1  1 15.04 14.96 13.17  9.29   NaN  9.87 13.67 10.25 10.83 12.58 18.50 15.04\n61  1  2 14.71   NaN 10.83  6.50 12.62  7.67 11.50 10.04  9.79  9.67 17.54 13.83\n61  1  3 18.50 16.88 12.33 10.13 11.17  6.17 11.25   NaN  8.50  7.67 12.75 12.71\n'

   The first three columns are year, month and day.  The
   remaining 12 columns are average windspeeds in knots at 12
   locations in Ireland on that day.   

   More information about the dataset go [here](wind.desc).

In [92]:
# Link does not work.

### Step 1. Import the necessary libraries

In [93]:
# Aha.

### Step 2. Import the dataset from this [address](https://raw.githubusercontent.com/guipsamora/pandas_exercises/master/06_Stats/Wind_Stats/wind.data)

In [111]:
data_url = "https://raw.githubusercontent.com/guipsamora/pandas_exercises/master/06_Stats/Wind_Stats/wind.data"

### Step 3. Assign it to a variable called data and replace the first 3 columns by a proper datetime index.

In [365]:
data = pd.read_csv(data_url, delim_whitespace=True)
data["Yr"] = pd.to_datetime(f"{data.Yr}-{data.Mo}-{data.Dy}" , format="%y-%m-%d")
# data.iloc[:, :3] = data.iloc[:, :3].apply(pd.to_datetime, format="%Y-%m-%d", errors="coerce")

data
# data["Yr"]

# x = pd.to_datetime(f"{61}-{2}-{13}" , format="%y-%m-%d")
# x

ValueError: ignored

### Step 4. Year 2061? Do we really have data from this year? Create a function to fix it and apply it.

In [260]:
# Turns out we needed an import! Tricks on me
import numpy as np

# Checkout this cool lambda. Checks if date exceeds current year (2022 at the time of coding), if true, substracts a 100 years
fix_date = lambda x: x - np.timedelta64(100,'Y') if x > pd.to_datetime("2022") else x

# Apply lambda and normalize date to leave only date, but keep the datetime type
data["Yr"] = data["Yr"].apply(fix_date).dt.normalize()

### Step 5. Set the right dates as the index. Pay attention at the data type, it should be datetime64[ns].

In [261]:
data.set_index("Yr")

Mo    Dy  Unnamed: 3  Unnamed: 4    RPT  Unnamed: 6  Unnamed: 7  \
Yr                                                                              
1961-01-01   NaN   1.0         NaN        1.00  15.04       14.96       13.17   
1961-01-01   NaN   1.0         NaN        2.00  14.71         NaN         NaN   
1961-01-01   NaN   1.0         NaN        3.00  18.50       16.88       12.33   
1961-01-01   NaN   1.0         NaN        4.00  10.58         NaN        6.63   
1961-01-01   NaN   1.0         NaN        5.00  13.33       13.25       11.42   
...          ...   ...         ...         ...    ...         ...         ...   
1978-01-01  12.0  27.0       17.58       16.96  17.62         NaN        8.08   
1978-01-01  12.0  28.0       13.21         NaN   5.46       13.46         NaN   
1978-01-01  12.0  29.0       14.00       10.29  14.42         NaN        8.71   
1978-01-01  12.0  30.0       18.50       14.04  21.29         NaN        9.13   
1978-01-01  12.0  31.0       20.33       17.41  27.29         NaN        9.59   

              VAL  Unnamed: 9  Unnamed: 10  ...  MUL  Unnamed: 30  \
Yr                                          ...                     
1961-01-01    NaN        9.29          NaN  ...  NaN          NaN   
1961-01-01    NaN       10.83          NaN  ...  NaN          NaN   
1961-01-01  10.13       11.17          NaN  ...  NaN          NaN   
1961-01-01  11.75         NaN         4.58  ...  NaN          NaN   
1961-01-01    NaN        6.17        10.71  ...  NaN          NaN   
...           ...         ...          ...  ...  ...          ...   
1978-01-01  13.21       11.67        14.46  ...  NaN          NaN   
1978-01-01   5.00         NaN         8.12  ...  NaN          NaN   
1978-01-01    NaN        9.71        10.54  ...  NaN          NaN   
1978-01-01  12.75         NaN         9.71  ...  NaN          NaN   
1978-01-01  12.08       10.13        19.25  ...  NaN          NaN   

            Unnamed: 31  CLO  Unnamed: 33  Unnamed: 34  BEL  Unnamed: 36  \
Yr                                                                         
1961-01-01          NaN  NaN          NaN          NaN  NaN          NaN   
1961-01-01          NaN  NaN          NaN          NaN  NaN          NaN   
1961-01-01          NaN  NaN          NaN          NaN  NaN          NaN   
1961-01-01          NaN  NaN          NaN          NaN  NaN          NaN   
1961-01-01          NaN  NaN          NaN          NaN  NaN          NaN   
...                 ...  ...          ...          ...  ...          ...   
1978-01-01          NaN  NaN          NaN          NaN  NaN          NaN   
1978-01-01          NaN  NaN          NaN          NaN  NaN          NaN   
1978-01-01          NaN  NaN          NaN          NaN  NaN          NaN   
1978-01-01          NaN  NaN          NaN          NaN  NaN          NaN   
1978-01-01          NaN  NaN          NaN          NaN  NaN          NaN   

            Unnamed: 37  MAL  
Yr                            
1961-01-01          NaN  NaN  
1961-01-01          NaN  NaN  
1961-01-01          NaN  NaN  
1961-01-01          NaN  NaN  
1961-01-01          NaN  NaN  
...                 ...  ...  
1978-01-01          NaN  NaN  
1978-01-01          NaN  NaN  
1978-01-01          NaN  NaN  
1978-01-01          NaN  NaN  
1978-01-01          NaN  NaN  

[6574 rows x 38 columns]

### Step 6. Compute how many values are missing for each location over the entire record.  
#### They should be ignored in all calculations below. 

In [291]:
non_location_columns = ("Yr", "Mo", "Dy", "Unnamed")
location_filter = ~data.columns.str.startswith(non_location_columns)
data.loc[:, location_filter].isnull().sum()

RPT    1625
VAL    2432
ROS    2152
KIL    2331
SHA    3204
BIR    4065
DUB    4416
CLA    5590
MUL    6574
CLO    6574
BEL    6574
MAL    6574
dtype: int64

### Step 7. Compute how many non-missing values there are in total.

In [293]:
# First sum counts for each column and seconds counts total
data.notnull().sum().sum()

98579

### Step 8. Calculate the mean windspeeds of the windspeeds over all the locations and all the times.
#### A single number for the entire dataset.

In [294]:
data

Yr    Mo    Dy  Unnamed: 3  Unnamed: 4    RPT  Unnamed: 6  \
0    1961-01-01   NaN   1.0         NaN        1.00  15.04       14.96   
1    1961-01-01   NaN   1.0         NaN        2.00  14.71         NaN   
2    1961-01-01   NaN   1.0         NaN        3.00  18.50       16.88   
3    1961-01-01   NaN   1.0         NaN        4.00  10.58         NaN   
4    1961-01-01   NaN   1.0         NaN        5.00  13.33       13.25   
...         ...   ...   ...         ...         ...    ...         ...   
6569 1978-01-01  12.0  27.0       17.58       16.96  17.62         NaN   
6570 1978-01-01  12.0  28.0       13.21         NaN   5.46       13.46   
6571 1978-01-01  12.0  29.0       14.00       10.29  14.42         NaN   
6572 1978-01-01  12.0  30.0       18.50       14.04  21.29         NaN   
6573 1978-01-01  12.0  31.0       20.33       17.41  27.29         NaN   

      Unnamed: 7    VAL  Unnamed: 9  ...  MUL  Unnamed: 30  Unnamed: 31  CLO  \
0          13.17    NaN        9.29  ...  NaN          NaN          NaN  NaN   
1            NaN    NaN       10.83  ...  NaN          NaN          NaN  NaN   
2          12.33  10.13       11.17  ...  NaN          NaN          NaN  NaN   
3           6.63  11.75         NaN  ...  NaN          NaN          NaN  NaN   
4          11.42    NaN        6.17  ...  NaN          NaN          NaN  NaN   
...          ...    ...         ...  ...  ...          ...          ...  ...   
6569        8.08  13.21       11.67  ...  NaN          NaN          NaN  NaN   
6570         NaN   5.00         NaN  ...  NaN          NaN          NaN  NaN   
6571        8.71    NaN        9.71  ...  NaN          NaN          NaN  NaN   
6572        9.13  12.75         NaN  ...  NaN          NaN          NaN  NaN   
6573        9.59  12.08       10.13  ...  NaN          NaN          NaN  NaN   

      Unnamed: 33  Unnamed: 34  BEL  Unnamed: 36  Unnamed: 37  MAL  
0             NaN          NaN  NaN          NaN          NaN  NaN  
1             NaN          NaN  NaN          NaN          NaN  NaN  
2             NaN          NaN  NaN          NaN          NaN  NaN  
3             NaN          NaN  NaN          NaN          NaN  NaN  
4             NaN          NaN  NaN          NaN          NaN  NaN  
...           ...          ...  ...          ...          ...  ...  
6569          NaN          NaN  NaN          NaN          NaN  NaN  
6570          NaN          NaN  NaN          NaN          NaN  NaN  
6571          NaN          NaN  NaN          NaN          NaN  NaN  
6572          NaN          NaN  NaN          NaN          NaN  NaN  
6573          NaN          NaN  NaN          NaN          NaN  NaN  

[6574 rows x 39 columns]

### Step 9. Create a DataFrame called loc_stats and calculate the min, max and mean windspeeds and standard deviations of the windspeeds at each location over all the days 

#### A different set of numbers for each location.

### Step 10. Create a DataFrame called day_stats and calculate the min, max and mean windspeed and standard deviations of the windspeeds across all the locations at each day.

#### A different set of numbers for each day.

### Step 11. Find the average windspeed in January for each location.  
#### Treat January 1961 and January 1962 both as January.

### Step 12. Downsample the record to a yearly frequency for each location.

### Step 13. Downsample the record to a monthly frequency for each location.

### Step 14. Downsample the record to a weekly frequency for each location.

### Step 15. Calculate the min, max and mean windspeeds and standard deviations of the windspeeds across all locations for each week (assume that the first week starts on January 2 1961) for the first 52 weeks.

## Section 7: Visualization

### Introduction:

This exercise is based on the titanic Disaster dataset avaiable at [Kaggle](https://www.kaggle.com/c/titanic).  
To know more about the variables check [here](https://www.kaggle.com/c/titanic/data)


### Step 1. Import the necessary libraries

### Step 2. Import the dataset from this [address](https://raw.githubusercontent.com/guipsamora/pandas_exercises/master/07_Visualization/Titanic_Desaster/train.csv)

### Step 3. Assign it to a variable titanic 

### Step 4. Set PassengerId as the index 

### Step 5. Create a pie chart presenting the male/female proportion

### Step 6. Create a scatterplot with the Fare payed and the Age, differ the plot color by gender

### Step 7. How many people survived?

### Step 8. Create a histogram with the Fare payed

### BONUS: Create your own question and answer it.

## Section 8: Creating Series and DataFrames

### Introduction:

This time you will create the data.



### Step 1. Import the necessary libraries

### Step 2. Create a data dictionary that looks like the DataFrame below

### Step 3. Assign it to a variable called pokemon

### Step 4. Ops...it seems the DataFrame columns are in alphabetical order. Place  the order of the columns as name, type, hp, evolution, pokedex

### Step 5. Add another column called place, and insert what you have in mind.

### Step 6. Present the type of each column

### BONUS: Create your own question and answer it.

## Section: 9 Time Series

### Introduction:

We are going to use Apple's stock price.


### Step 1. Import the necessary libraries

### Step 2. Import the dataset from this [address](https://raw.githubusercontent.com/guipsamora/pandas_exercises/master/09_Time_Series/Apple_Stock/appl_1980_2014.csv)

### Step 3. Assign it to a variable apple

### Step 4.  Check out the type of the columns

### Step 5. Transform the Date column as a datetime type

### Step 6.  Set the date as the index

### Step 7.  Is there any duplicate dates?

### Step 8.  Ops...it seems the index is from the most recent date. Make the first entry the oldest date.

### Step 9. Get the last business day of each month

### Step 10.  What is the difference in days between the first day and the oldest

### Step 11.  How many months in the data we have?

### Step 12. Plot the 'Adj Close' value. Set the size of the figure to 13.5 x 9 inches

## Section 10: Deleting

### Introduction:

This exercise is a adaptation from the UCI Wine dataset.
The only pupose is to practice deleting data with pandas.

### Step 1. Import the necessary libraries

### Step 2. Import the dataset from this [address](https://archive.ics.uci.edu/ml/machine-learning-databases/wine/wine.data). 

### Step 3. Assign it to a variable called wine

### Step 4. Delete the first, fourth, seventh, nineth, eleventh, thirteenth and fourteenth columns

### Step 5. Assign the columns as below:

The attributes are (donated by Riccardo Leardi, riclea '@' anchem.unige.it):  
1) alcohol  
2) malic_acid  
3) alcalinity_of_ash  
4) magnesium  
5) flavanoids  
6) proanthocyanins  
7) hue 

### Step 6. Set the values of the first 3 rows from alcohol as NaN

### Step 7. Now set the value of the rows 3 and 4 of magnesium as NaN

### Step 8. Fill the value of NaN with the number 10 in alcohol and 100 in magnesium

### Step 9. Count the number of missing values

### Step 10.  Create an array of 10 random numbers up until 10

### Step 11.  Use random numbers you generated as an index and assign NaN value to each of cell.

### Step 12.  How many missing values do we have?

### Step 13. Delete the rows that contain missing values

### Step 14. Print only the non-null values in alcohol

### Step 15.  Reset the index, so it starts with 0 again